tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/signatures/ApalacheOperSignatures.scala

```diff
@@ -0,0 +1,75 @@
+package at.forsyte.apalache.tla.typecomp.signatures
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.oper.ApalacheOper
+import at.forsyte.apalache.tla.typecomp.{BuilderUtil, SignatureMap}
+
+/**
+ * Produces a SignatureMap for all internal Apalache operators
+ *
+ * @author
+ *   Jure Kukovec
+ */
+object ApalacheOperSignatures {
+  import ApalacheOper._
+  import BuilderUtil._
+
+  def getMap: SignatureMap = {
+
+    // (t,t) => Bool
+    val assignSig = signatureMapEntry(assign, { case Seq(t, tt) if t == tt => BoolT1 })
+
+    // gen has no signature, because we can't encode
+    // (Int) => t
+
+    // (Bool) => Bool
+    val skolemSig = signatureMapEntry(skolem, { case Seq(BoolT1) => BoolT1 })
+
+    // (Set(t)) => t
+    val guessSig = signatureMapEntry(guess, { case Seq(SetT1(t)) => t })
+
+    // (Set(Set(t))) => Set(Set(t))
+    // or
+    // (Set(a -> b)) => Set(a -> b)
+    val expandSig = signatureMapEntry(expand,
+        {
+          case Seq(powsetT @ SetT1(_: SetT1)) => powsetT
+          case Seq(funsetT @ SetT1(_: FunT1)) => funsetT
+        })
+
+    // (Bool) => Bool
+    val constCardSig = signatureMapEntry(constCard, { case Seq(BoolT1) => BoolT1 })
+
+    // funAsSeq should not be constructed
+
+    // (Int, Int => t) => Seq(t)
+    val mkSeqSig = signatureMapEntry(mkSeq, { case Seq(IntT1, OperT1(Seq(t), IntT1)) => SeqT1(t) })
+
+    // ((t1,t2) => t1, t1, Set(t2)) => t1
+    val foldSetSig = signatureMapEntry(foldSet,
+        {
+          case Seq(OperT1(Seq(t1, t2), t11), t12, SetT1(t22)) if t1 == t11 && t11 == t12 && t2 == t22 => t1
+        })
+
+    // ((t1,t2) => t1, t1, Seq(t2)) => t1
+    val foldSeqSig = signatureMapEntry(foldSeq,
+        {
+          case Seq(OperT1(Seq(t1, t2), t11), t12, SeqT1(t22)) if t1 == t11 && t11 == t12 && t2 == t22 => t1
```

Comment:
quite hard for me to work out which types are meant to go with which. Maybe this would be better?

```suggestion
    // ((a,b) => a, a, Seq(b)) => a
    val foldSeqSig = signatureMapEntry(foldSeq,
        {
          case Seq(OperT1(Seq(a1, b1), a2), a3, SeqT1(b2)) if a1 == a2 && a2 == a3 && b1 == b2 => a1
```

The idea being that all the `a`s and `b`s should be equal.

Reply:
Makes sense, I'll change the two sigs.